File: iOS/AUIKitCore/Sources/Service/Arbiter/AUIArbiter.swift
//
//  AUIArbiter.swift
//  AUIKitCore
//
//  Created by wushengtao on 2023/11/2.
//

import AgoraRtmKit

@objc public class AUIArbiter: NSObject {
    private var channelName: String!
    private var rtmManager: AUIRtmManager!
    private var currentUserInfo: AUIUserThumbnailInfo!
    private(set) var lockOwnerId: String = ""
    
    deinit {
        aui_info("deinit AUIArbiter", tag: "AUIArbiter")
        rtmManager.unsubscribeLock(channelName: channelName, lockName: kRTM_Referee_LockName, delegate: self)
    }
    
    public init(channelName: String, rtmManager: AUIRtmManager, userInfo: AUIUserThumbnailInfo) {
        self.rtmManager = rtmManager
        self.channelName = channelName
        self.currentUserInfo = userInfo
        super.init()
        rtmManager.subscribeLock(channelName: channelName, lockName: kRTM_Referee_LockName, delegate: self)
        aui_info("init AUIArbiter", tag: "AUIArbiter")
    }
    
    public func create() {
        rtmManager.setLock(channelName: channelName, lockName: kRTM_Referee_LockName) { err in
        }
    }
    
    public func destroy() {
        rtmManager.removeLock(channelName: channelName, lockName: kRTM_Referee_LockName) { err in
        }
    }
    
    public func acquire() {
        rtmManager.acquireLock(channelName: channelName, lockName: kRTM_Referee_LockName) { err in
        }
    }
    
    public func release() {
        rtmManager.releaseLock(channelName: channelName, lockName: kRTM_Referee_LockName) { err in
        }
    }
    
    
    public func isArbiter() -> Bool {
        return lockOwnerId == currentUserInfo.userId
    }
}

//MARK: AUIRtmLockProxyDelegate
extension AUIArbiter: AUIRtmLockProxyDelegate {
    public func onReceiveLockDetail(channelName: String, lockDetail: AgoraRtmLockDetail) {
        aui_info("onReceiveLockDetail[\(channelName)]: \(lockDetail.owner)/\(currentUserInfo.userId)")
        guard channelName == self.channelName else {return}
        
        //如果lockOwnerId是自己，并且是切换了仲裁者(非首次获取，否则第一次roomService里onReceiveLockDetail拿到的是空)，需要刷新下
        if lockOwnerId.isEmpty == false, lockDetail.owner == currentUserInfo.userId {
            rtmManager.fetchMetaDataSnapshot(channelName: channelName) {[weak self] error in
                //TODO: error handler, retry?
                self?.lockOwnerId = lockDetail.owner
            }
        } else {
            lockOwnerId = lockDetail.owner
        }
    }
    
    public func onReleaseLockDetail(channelName: String, lockDetail: AgoraRtmLockDetail) {
        aui_info("onReleaseLockDetail[\(channelName)]: \(lockDetail.owner)")
        guard channelName == self.channelName else {return}
        rtmManager.acquireLock(channelName: channelName, lockName: kRTM_Referee_LockName) { err in
        }
    }
}

File: Android/auikit-service/src/main/java/io/agora/auikit/service/arbiter/AUIArbiter.kt
package io.agora.auikit.service.arbiter

import io.agora.auikit.service.callback.AUIException
import io.agora.auikit.service.rtm.AUIRtmLockRespObserver
import io.agora.auikit.service.rtm.AUIRtmManager

class AUIArbiter(
    private val channelName: String,
    private val rtmManager: AUIRtmManager,
    private val currentUserId: String
) {

    private var lockOwnerId = ""
    private val rtmLockRespObserver = object : AUIRtmLockRespObserver {
        override fun onReceiveLock(channelName: String, lockName: String, lockOwner: String) {
            if (lockOwnerId.isNotEmpty() && lockOwner == currentUserId) {
                rtmManager.fetchMetaDataSnapshot(channelName) {
                    //TODO: error handler, retry?
                    lockOwnerId = lockOwner
                }
            } else {
                lockOwnerId = lockOwner
            }
        }

        override fun onReleaseLock(channelName: String, lockName: String, lockOwner: String) {
            if (channelName == this@AUIArbiter.channelName) {
                acquire()
            }
        }
    }

    init {
        rtmManager.subscribeLock(channelName, observer = rtmLockRespObserver)
    }

    fun deInit(){
        rtmManager.unsubscribeLock(rtmLockRespObserver)
    }

    fun create(){
        rtmManager.setLock(channelName){}
    }

    fun destroy(){
        rtmManager.removeLock(channelName){}
    }

    fun acquire(callback: ((AUIException?)->Unit)? = null) {
        rtmManager.acquireLock(channelName){
            callback?.invoke(if(it == null) null else AUIException(AUIException.ERROR_CODE_RTM, "$it"))
        }
    }

    fun release() {
        rtmManager.releaseLock(channelName){}
    }

    fun isArbiter() = lockOwnerId == currentUserId

    fun lockOwnerId() = lockOwnerId
}